src/lib/core/event.h

```diff
@@ -0,0 +1,251 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "func_adapter.h"
+#include "small/rlist.h"
+#include "trivia/util.h"
+#include <stdbool.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/**
+ * A node of the list, containing func_adapter.
+ * Since event_triggers have completely different call interface,
+ * they are not inherited from core triggers.
+ */
+struct event_trigger {
+	/** A link in a list of all registered event triggers. */
+	struct rlist link;
+	/** Trigger function. */
+	struct func_adapter *func;
+	/** Backlink to event. Needed for reference count of event. */
+	struct event *event;
+	/** Unique name of the trigger. */
+	char *name;
+	/** Trigger reference counter. */
+	uint32_t ref_count;
+	/** This flag is set when the trigger is deleted. */
+	bool is_deleted;
+};
+
+/**
+ * Creates trigger from func_adapter and name. Created trigger owns the adapter
+ * and will destroy it in its destructor. Passed name should be zero terminated
+ * and will be copied.
+ */
+struct event_trigger *
+event_trigger_new(struct func_adapter *func, const char *name);
+
+/**
+ * Destroys an event_trigger. Underlying func_adapter is destroyed,
+ * event_trigger is deallocated.
```

Comment:
Nit: I would mention that this function is for internal use only and that `event_trigger_unref` must be used.

Reply:
Added the comment by myself.